### DISCUSSION TITLE
Select/unselect tracks from CLI

### DISCUSSION THREAD

**Discussion by weber8thomas** [GMOD/jbrowse-components#4153] [Q&A]
Hi everyone,

Is there a way to select/unselect tracks programatically from CLI? I'm currently building a use case that contains hundreds/thousands of tracks, and change dynamically the list of tracks selected using the CLI based on some external filtering

Many thanks in advance,

Cheers, Thomas

**Comment by cmdcolin** [comment]
is this question sort of related to controlling a running instance of jbrowse or something different?

**Reply by weber8thomas** [reply]
I'd like to control a running instance via the CLI yes!

**Reply by cmdcolin** [reply]
this is a sort of interesting idea. there are a number of different jbrowse 2 "products" so it could depend on how you want to approach. for example, there is the "jbrowse-web" product, which is the basic static website instance you get from e.g. "jbrowse create yourfolder". there is "jbrowse-desktop" that is the electron desktop instance. then there are also the NPM libraries, and the things like jbrowse-jupyter and JBrowseR. 

the NPM libraries (@jbrowse/react-linear-genome-view, @jbrowse/react-app, etc https://jbrowse.org/jb2/docs/embedded_components/) offer the most control, since you can use them programmatically, so might be a good option for your use case.  you can run actions/functions/methods on the running instance like addTrack and hideTrack, etc

the jbrowse-jupyter also works well in a notebook environment, so it's not CLI necessarily, but is sort of similarly programmable like the NPM packages. we are working on improving the notebook usages lately.

then there is jbrowse-web and jbrowse-desktop, and generally these are programmable via creating custom plugins. you can obtain a similar level of control that you get with the NPM libraries, but you just do it from within a plugin.

in any case, this would probably involve some custom code, but hope that sort of outlines the options that are available. if you wanted to target any specific one of those routes, i can elaborate more!

**Reply by weber8thomas** [reply]
Thanks @cmdcolin for your answer! 

I'll also try to elaborate a bit more in order to give you more context. I'm currently developing a plotly dash platform that allows people to design themself their dashboard by designing and laying out generic plots (violin/boxplot/scatter ...), metrics card, interactive components (dropdown, slider, text input) as well as adding a jbrowse instance on the dashboard. My idea was that the interactivity provided by the dropdowns/slider should dynamically update both the plots and metrics cards as well as the list of tracks displayed on the jbrowse instance. I started to use the jbrowse-dash flavour  which seemed to be the most straightforward option but encountered some limitations regarding tracks coloring using jexl, as well as kind of a "reset" where all the selected tracks were removed and the genomic position set back to default. I also tried to containerise and use the jbrowse-web version, packing it as a iframe in my dash app, and then experimenting with the CLI to add/update/remove tracks. 

In summary, what I'd need would be: a way to add/remove tracks to the global list, add/remove tracks to the display programatically, a way to access the current list of tracks displayed as well as the current genomic position selected. 

What would you advise to do so? I'm fully open to help contributing here! :)

**Reply by cmdcolin** [reply]
that's interesting, i'm glad you tried out dash_jbrowse but indeed i think the challenges you see are interesting too, and it's something we want to improve on (similar feature request: https://github.com/GMOD/jbrowse-jupyter/issues/86 related to controlling the instance "after launch" so you don't get the thing where the browser fully resets)

we (or you!) could try to add some more work on dash_jbrowse to enable your use case, or alternatively, you can skip dash_jbrowse and incorporate e.g. @jbrowse/react-linear-genome-view into your own dash component (there is not very much  code in dash_jbrowse and it's just a small wrapper, but that wrapper could even be getting in the way of your work, so just using the plain NPM package without the dash_jbrowse wrapper could potentially be helpful)

**Reply by weber8thomas** [reply]
Hi @cmdcolin, thanks for your answer! I got a second look at the docs and noticed you already have a system in place to partially solves my problem with: https://jbrowse.org/jb2/docs/urlparams/ where I can directly give the assembly + location, the list of tracks and additional parameters there to configure the browser. As currently everything is directly stored into a local storage and the URL is simplified, is there a way to retrieve the modification done by the user on the UI (change of location, tracks enabled/disabled), using the local storage data, or something else? Many thanks

**Reply by cmdcolin** [reply]
I am not sure exactly what type of workflow would solve it, but you could make a plugin that 'watches' for these changes

example

```
// basic plugin that uses an autorun to detect when tracks change and view location changes
;(function () {
  class Plugin {
    name = 'WatcherPlugin'
    version = '1.0'

    install(/* pluginManager */) {
      // do nothing
    }

    configure(pluginManager) {
      const { autorun } = pluginManager.jbrequire('mobx')
      autorun(() => {
        if (pluginManager.rootModel?.session) {
          // console.log'ing the trackIds that are open and the 
          console.log(
            pluginManager.rootModel.session.views[0].tracks.map(
              t => t.configuration.trackId,
            ),
          )
          // this is the current visible location in the view e.g. chr1:1..100
          console.log(
            pluginManager.rootModel.session.views[0].coarseVisibleLocStrings,
          )
        }
      })
    }
  }
  ;(typeof self !== 'undefined' ? self : window).JBrowsePluginWatcherPlugin = {
    default: Plugin,
  }
})()
```

you could change the console.logs to make an api request, or write to localstorage, or something like that

**Reply by weber8thomas** [reply]
Hi @cmdcolin , I developed [this plugin](https://github.com/weber8thomas/jbrowse-watcher-plugin/blob/main/src/WatcherPlugin/index.ts) that allows both to listen track changes and location change and send the data to an API in order to use it for my system. I added a timeout to wait every second for changes and prevent sending multiple batch of POST when changing and adjusting location. I didn't find a way however to configure properly and cleanly the API hostname/port/endpoint to target if wished. Any thoughts on this ?

**Reply by cmdcolin** [reply]
great work. currently plugins can register some "global config" like this example from LinearGenomeView  https://github.com/GMOD/jbrowse-components/blob/ca55dbd21d6c742471a4a8f7a04d1fd1863322f3/plugins/linear-genome-view/src/index.ts#L39-L56

this configuration that is registered onto the session model, so you can read configuration off the session like this

```
getConf(session, [ 'LinearGenomeViewPlugin',  'trackLabels' ])
```

so the syntax is getConf(session, [ 'YourPluginName',  'configSlotName' ])

and then an example of this being configured in the config.json is like this

https://github.com/GMOD/jbrowse-components/blob/ca55dbd21d6c742471a4a8f7a04d1fd1863322f3/test_data/volvox/config.json#L189-L192
